File: internal/agui/studio_api.go
package agui

// studio_api.go serves the cockpit Studio: an identity generates images and videos from one
// composer bar, outside any conversation, and reads its own history and library. Every route
// is scoped to the authenticated principal — the owner is never taken from the body — and
// every generation goes through the same shared paths the agent's tools use.

import (
	"context"
	"net/http"
	"strconv"

	"github.com/chetto1983/aura/internal/assets"
	"github.com/chetto1983/aura/internal/mediagen"
)

// studioHistoryDefault is one screen of history; the ceiling is the store's own page cap.
const studioHistoryDefault = 24

// StudioBackend is the Studio's live half: the catalog it lists, the two shared generation
// paths it pays through, the job store it reads back, and the identity's image library. The
// composition root implements it over the daemon's media dependencies (cmd/aura), so this
// package needs neither a provider client nor a database.
type StudioBackend interface {
	Models(ctx context.Context, kind mediagen.Kind) (defaultModel string, models []mediagen.Model, err error)
	SubmitVideo(ctx context.Context, owner string, req StudioVideoRequest) (mediagen.Job, error)
	GenerateImage(ctx context.Context, owner string, req StudioImageRequest) (mediagen.Job, error)
	History(ctx context.Context, owner, beforeID string, kind mediagen.Kind, limit int) ([]mediagen.Job, error)
	Library(ctx context.Context, owner string, limit int) ([]assets.Asset, error)
	FinalizeUpload(ctx context.Context, owner, assetID string) (assets.Asset, error)
}

func (s *Server) registerStudioRoutes(mux *http.ServeMux) {
	mux.HandleFunc("GET /api/studio/models", s.handleStudioModels)
	mux.HandleFunc("POST /api/studio/videos", s.handleStudioVideoCreate)
	mux.HandleFunc("POST /api/studio/images", s.handleStudioImageCreate)
	mux.HandleFunc("GET /api/studio/history", s.handleStudioHistory)
	mux.HandleFunc("GET /api/studio/library", s.handleStudioLibrary)
	mux.HandleFunc("POST /api/studio/uploads/{id}/finalize", s.handleStudioUploadFinalize)
}

// studioCaller is the gate every Studio route shares: an unwired Studio is unavailable, and
// an unauthenticated caller has no identity to own a generation.
func (s *Server) studioCaller(w http.ResponseWriter, r *http.Request) (string, bool) {
	if s.studio == nil {
		http.Error(w, "studio unavailable", http.StatusServiceUnavailable)
		return "", false
	}
	identityID, ok := principalIdentityID(r)
	if !ok {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return "", false
	}
	return identityID, true
}

// studioKind reads a kind from a query value. Only the two kinds the catalog knows are
// accepted, so a stale or forged form naming a third is refused rather than passed on.
func studioKind(raw string) (mediagen.Kind, bool) {
	switch kind := mediagen.Kind(raw); kind {
	case mediagen.KindImage, mediagen.KindVideo:
		return kind, true
	}
	return "", false
}

// studioLimit reads a page size: absent means the default, anything the caller names is
// clamped to [1, ceiling], and a value that is not a number is a refusal rather than a
// silent fallback — a cockpit sending one has a bug the operator should see.
func studioLimit(raw string, fallback, ceiling int) (int, bool) {
	if raw == "" {
		return fallback, true
	}
	limit, err := strconv.Atoi(raw)
	if err != nil {
		return 0, false
	}
	return min(max(limit, 1), ceiling), true
}

func (s *Server) handleStudioModels(w http.ResponseWriter, r *http.Request) {
	_, ok := s.studioCaller(w, r)
	if !ok {
		return
	}
	kind, ok := studioKind(r.URL.Query().Get("kind"))
	if !ok {
		http.Error(w, "kind must be image or video", http.StatusBadRequest)
		return
	}
	defaultModel, models, err := s.studio.Models(r.Context(), kind)
	if err != nil {
		writeStudioError(w, err)
		return
	}
	out := studioModelsDTO{Default: defaultModel, Models: make([]studioModelDTO, 0, len(models))}
	for _, model := range models {
		out.Models = append(out.Models, studioModel(model, kind))
	}
	writeJSON(w, out)
}

// studioCreate is what both create routes are: the principal owns the generation, the body is
// decoded strictly so an unknown field is a refusal rather than a silently dropped intent, and
// the accepted record is the answer. create is a closure, not a method value, because the
// backend is only known to be present once studioCaller has passed.
func studioCreate[T any](s *Server, w http.ResponseWriter, r *http.Request,
	create func(context.Context, string, T) (mediagen.Job, error),
) {
	identityID, ok := s.studioCaller(w, r)
	if !ok {
		return
	}
	var body T
	if err := strictDecodeJSON(w, r, &body, decodeOpts{}); err != nil {
		http.Error(w, "invalid request body", http.StatusBadRequest)
		return
	}
	job, err := create(r.Context(), identityID, body)
	if err != nil {
		writeStudioError(w, err)
		return
	}
	writeJSONStatus(w, http.StatusCreated, studioRecord(job))
}

func (s *Server) handleStudioVideoCreate(w http.ResponseWriter, r *http.Request) {
	studioCreate(s, w, r, func(ctx context.Context, owner string, body StudioVideoRequest) (mediagen.Job, error) {
		return s.studio.SubmitVideo(ctx, owner, body)
	})
}

func (s *Server) handleStudioImageCreate(w http.ResponseWriter, r *http.Request) {
	studioCreate(s, w, r, func(ctx context.Context, owner string, body StudioImageRequest) (mediagen.Job, error) {
		return s.studio.GenerateImage(ctx, owner, body)
	})
}

func (s *Server) handleStudioHistory(w http.ResponseWriter, r *http.Request) {
	identityID, ok := s.studioCaller(w, r)
	if !ok {
		return
	}
	query := r.URL.Query()
	var kind mediagen.Kind
	if raw := query.Get("kind"); raw != "" {
		if kind, ok = studioKind(raw); !ok {
			http.Error(w, "kind must be image or video", http.StatusBadRequest)
			return
		}
	}
	limit, ok := studioLimit(query.Get("limit"), studioHistoryDefault, mediagen.StudioPageMax)
	if !ok {
		http.Error(w, "limit must be a number", http.StatusBadRequest)
		return
	}
	jobs, err := s.studio.History(r.Context(), identityID, query.Get("before"), kind, limit)
	if err != nil {
		writeStudioError(w, err)
		return
	}
	out := studioHistoryDTO{Records: make([]studioRecordDTO, 0, len(jobs))}
	for _, job := range jobs {
		out.Records = append(out.Records, studioRecord(job))
	}
	writeJSON(w, out)
}

func (s *Server) handleStudioLibrary(w http.ResponseWriter, r *http.Request) {
	identityID, ok := s.studioCaller(w, r)
	if !ok {
		return
	}
	limit, ok := studioLimit(r.URL.Query().Get("limit"), studioHistoryDefault, mediagen.StudioPageMax)
	if !ok {
		http.Error(w, "limit must be a number", http.StatusBadRequest)
		return
	}
	library, err := s.studio.Library(r.Context(), identityID, limit)
	if err != nil {
		writeStudioError(w, err)
		return
	}
	out := studioLibraryDTO{Assets: make([]studioAssetDTO, 0, len(library))}
	for _, asset := range library {
		out.Assets = append(out.Assets, studioAsset(asset))
	}
	writeJSON(w, out)
}

func (s *Server) handleStudioUploadFinalize(w http.ResponseWriter, r *http.Request) {
	identityID, ok := s.studioCaller(w, r)
	if !ok {
		return
	}
	asset, err := s.studio.FinalizeUpload(r.Context(), identityID, r.PathValue("id"))
	if err != nil {
		writeStudioError(w, err)
		return
	}
	writeJSON(w, studioAsset(asset))
}

File: web/src/studio/studioApi.ts
// studioApi.ts — the browser half of internal/agui/studio_api.go. Every type below mirrors a
// `json:` tag in internal/agui/studio_dto.go: a field the Go DTO marks `omitempty` is optional
// here, because an absent capability or price is unknown and must never read as zero or free.

/** The two kinds the catalog knows; the routes refuse a third. */
export type StudioKind = 'image' | 'video';

/** One cell of a video model's price matrix: what a second costs at this resolution with this
 *  audio choice. The composer prices the clip being paid for, so a single rate would be wrong
 *  for every other cell. */
export interface StudioPrice {
  readonly resolution: string;
  readonly audio: boolean;
  readonly usd_per_second: number;
}

/** One picker row (studioModelDTO). `audio` and `seed` are always present — false means the
 *  model declares neither, which is also what an image row reads. */
export interface StudioModel {
  readonly id: string;
  readonly name?: string;
  readonly description?: string;
  readonly durations?: readonly number[];
  readonly resolutions?: readonly string[];
  readonly aspect_ratios?: readonly string[];
  /** 'first_frame' and/or 'last_frame' — the frames this model accepts. */
  readonly frame_images?: readonly string[];
  readonly audio: boolean;
  readonly seed: boolean;
  readonly prices?: readonly StudioPrice[];
  readonly reference_max?: number;
  readonly image_min_usd?: number;
  readonly image_max_usd?: number;
  readonly image_token_min_per_1m?: number;
  readonly image_token_max_per_1m?: number;
}

export interface StudioModels {
  readonly default: string;
  readonly models: readonly StudioModel[];
}

/** What the provider was actually asked for, after the server's clamp. */
export interface StudioUsed {
  readonly duration?: number;
  readonly resolution?: string;
  readonly aspect_ratio?: string;
  readonly audio?: boolean;
  readonly seed?: number;
  readonly first_frame_asset_id?: string;
  readonly last_frame_asset_id?: string;
  readonly reference_asset_ids?: readonly string[];
}

/** One history row, and the answer to both create routes. */
export interface StudioRecord {
  readonly id: string;
  readonly kind: StudioKind;
  readonly status: string;
  readonly model: string;
  readonly prompt: string;
  readonly used: StudioUsed;
  readonly adjustments?: readonly string[];
  /** Absent while the cost is not known yet — never 0. */
  readonly cost_usd?: number;
  readonly asset_id?: string;
  readonly error?: { readonly code: string; readonly message: string };
  readonly created_at: string;
  readonly completed_at?: string;
}

/** One library row. The server answers more (modality, status, size, created_at); the composer
 *  only ever needs to name the image it is attaching. */
export interface StudioImageRef {
  readonly id: string;
  readonly file_name: string;
  readonly mime_type: string;
}

/** POST /api/studio/videos' body. The route strict-decodes, so an undeclared field is a 400 —
 *  which is why every optional axis is omitted rather than sent as null. */
export interface StudioVideoBody {
  readonly model: string;
  readonly prompt: string;
  readonly duration?: number;
  readonly resolution?: string;
  readonly aspect_ratio?: string;
  readonly audio?: boolean;
  readonly seed?: number;
  readonly first_frame_asset_id?: string;
  readonly last_frame_asset_id?: string;
}

/** POST /api/studio/images' body. */
export interface StudioImageBody {
  readonly model: string;
  readonly prompt: string;
  readonly aspect_ratio?: string;
  readonly reference_asset_ids?: readonly string[];
}

/** A refusal the operator can act on: the status the route chose and the code + sentence the
 *  server wrote. An infrastructure failure has no code, so `code` is empty and the page falls
 *  back to its generic sentence rather than inventing one. */
export class StudioError extends Error {
  readonly status: number;
  readonly code: string;

  constructor(status: number, code: string, message: string) {
    super(message);
    this.name = 'StudioError';
    this.status = status;
    this.code = code;
  }
}

/** One screen of history — the server's own default (studioHistoryDefault), named so a caller
 *  can tell a full page from the last one. */
export const STUDIO_HISTORY_LIMIT = 24;

function readInit(signal?: AbortSignal): RequestInit {
  const init: RequestInit = {
    method: 'GET',
    headers: { Accept: 'application/json' },
    credentials: 'same-origin',
  };
  if (signal !== undefined) init.signal = signal;
  return init;
}

function writeInit(body?: unknown): RequestInit {
  const headers: Record<string, string> = { Accept: 'application/json' };
  if (body !== undefined) headers['Content-Type'] = 'application/json';
  const init: RequestInit = { method: 'POST', headers, credentials: 'same-origin' };
  if (body !== undefined) init.body = JSON.stringify(body);
  return init;
}

/** Read a Studio response, turning a failure into a StudioError.
 *
 *  A refused generation answers studioErrorDTO — `{code, error}`, the code in `code` and the
 *  sentence in `error`. The gate routes (unwired Studio, no principal, a body the route cannot
 *  decode) answer http.Error text instead, so a body that is not that object leaves the bare
 *  status and an empty code. */
async function studioJSON<T>(res: Response): Promise<T> {
  if (res.ok) return (await res.json()) as T;
  const fallback = `HTTP ${String(res.status)}`;
  const raw = await res.text().catch(() => '');
  let code = '';
  let message = fallback;
  try {
    const parsed: unknown = JSON.parse(raw);
    if (parsed !== null && typeof parsed === 'object') {
      const body = parsed as { readonly code?: unknown; readonly error?: unknown };
      if (typeof body.code === 'string') code = body.code;
      if (typeof body.error === 'string' && body.error.length > 0) message = body.error;
    }
  } catch {
    // Not JSON: the route answered plain text, and the status is all there is to report.
  }
  throw new StudioError(res.status, code, message);
}

export async function fetchStudioModels(
  kind: StudioKind,
  signal?: AbortSignal,
): Promise<StudioModels> {
  const res = await fetch(`/api/studio/models?kind=${kind}`, readInit(signal));
  return studioJSON<StudioModels>(res);
}

export async function listStudioHistory(
  kind: StudioKind | undefined,
  before: string | undefined,
  signal?: AbortSignal,
): Promise<readonly StudioRecord[]> {
  const query = new URLSearchParams({ limit: String(STUDIO_HISTORY_LIMIT) });
  if (kind !== undefined) query.set('kind', kind);
  if (before !== undefined && before.length > 0) query.set('before', before);
  const res = await fetch(`/api/studio/history?${query.toString()}`, readInit(signal));
  const page = await studioJSON<{ readonly records?: readonly StudioRecord[] }>(res);
  return page.records ?? [];
}

export async function createStudioVideo(body: StudioVideoBody): Promise<StudioRecord> {
  const res = await fetch('/api/studio/videos', writeInit(body));
  return studioJSON<StudioRecord>(res);
}

export async function createStudioImage(body: StudioImageBody): Promise<StudioRecord> {
  const res = await fetch('/api/studio/images', writeInit(body));
  return studioJSON<StudioRecord>(res);
}

export async function listStudioLibrary(signal?: AbortSignal): Promise<readonly StudioImageRef[]> {
  const res = await fetch('/api/studio/library', readInit(signal));
  const page = await studioJSON<{ readonly assets?: readonly StudioImageRef[] }>(res);
  return page.assets ?? [];
}

export async function finalizeStudioUpload(id: string): Promise<StudioImageRef> {
  const res = await fetch(`/api/studio/uploads/${encodeURIComponent(id)}/finalize`, writeInit());
  return studioJSON<StudioImageRef>(res);
}

/** The owned, identity-scoped asset routes. The Studio never learns an object key: a result is
 *  read back through the same gate the chat attachments use. */
export function assetDownloadUrl(id: string): string {
  return `/api/assets/${encodeURIComponent(id)}/download`;
}

export function assetStreamUrl(id: string): string {
  return `/api/assets/${encodeURIComponent(id)}/stream`;
}
